src/app/cmd/flatten_layers.cpp

```diff
@@ -97,44 +154,94 @@ void FlattenLayers::onExecute()
     RestoreVisibleLayers restore;
     restore.showSelectedLayers(sprite, layers);
 
+    // Map canvas to extended draw area
+    const gfx::ClipF area(
+      0, 0,
+      -area_x, -area_y,
+      spec.width(), spec.height());
+
     // Copy all frames to the background.
     for (frame_t frame(0); frame<sprite->totalFrames(); ++frame) {
       // Clear the image and render this frame.
       clear_image(image.get(), bgcolor);
-      render.renderSprite(image.get(), sprite, frame);
+      render.renderSprite(image.get(), sprite, frame, area);
 
-      // TODO Keep cel links when possible
+      // Get exact bounds for rendered frame
+      gfx::Rect bounds = image->bounds();
+      const bool shrink = doc::algorithm::shrink_bounds(
+        image.get(), image->maskColor(), nullptr,
+        image->bounds(), bounds);
 
-      ImageRef cel_image;
+      // Skip when fully transparent
       Cel* cel = flatLayer->cel(frame);
+      if (!shrink) {
+        if (!newFlatLayer && cel)
+          executeAndAdd(new cmd::RemoveCel(cel));
+
+        continue;
+      }
+
+      // Apply shrunk bounds to new image
+      const ImageRef new_image(doc::crop_image(
+        image.get(), bounds, image->maskColor()));
+
+      // Replace image on existing cel
       if (cel) {
+
+        // TODO Keep cel links when possible
+
         if (cel->links())
           executeAndAdd(new cmd::UnlinkCel(cel));
 
-        cel_image = cel->imageRef();
+        const ImageRef cel_image = cel->imageRef();
         ASSERT(cel_image);
 
-        executeAndAdd(
-          new cmd::CopyRect(cel_image.get(), image.get(),
-                            gfx::Clip(0, 0, image->bounds())));
+        // Reset cel properties when flattening in-place
+        if (!newFlatLayer) {
+          executeAndAdd(new cmd::SetCelOpacity(cel, 255));
+          executeAndAdd(new cmd::SetCelPosition(cel,
+            area.src.x+bounds.x, area.src.y+bounds.y));
+        }
+
+        // Modify destination cel
+        executeAndAdd(new cmd::ReplaceImage(sprite, cel_image, new_image));
```

Comment:
Just as a side comment: Something for the future (not for this PR, to keep it as simple as possible), here we could do something like in [FilterManagerImpl::apply()](https://github.com/aseprite/aseprite/blob/81bf86f9e700ca2e75434256a986edf1f4388297/src/app/commands/filters/filter_manager_impl.cpp#L223) using PatchCel/CopyRegion for certain cases (to reduce the undo information). But this is not a priority.